EMF2014/RadioTask.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ TiLDA Mk2
+ 
+ RadioTask
+ 
+ The MIT License (MIT)
+ 
+ Copyright (c) 2014 Electromagnetic Field LTD
+ 
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in all
+ copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ */
+
+#include "RadioTask.h"
+#include "Tilda.h"
+#include "Utils.h"
+#include "Deserializer.h"
+#include <debug.h>
+
+RadioTask::RadioTask() {
+    _messages = xQueueCreate(RADIO_MAX_INCOMING_MESSAGES_BUFFER, sizeof(Serializable *));
+    _wakeUpSignal = xQueueCreate(1, sizeof(byte));
+    pinMode(PIN_LED_RXL, OUTPUT);
+    _listening = false;
+}
+
+String RadioTask::getName() const {
+    return "RadioTask";
+}
+
+void RadioTask::listen(uint8_t channel, uint16_t panId) {
+    _channel = channel;
+    _panId = panId;
+    _changed = true;
+    _listening = true;
+    xQueueReset(_messages);
+    byte meaningless = 1;
+    xQueueOverwrite(_wakeUpSignal, &meaningless);
+}
+
+void RadioTask::close() {
+    _listening = false;
+    xQueueReset(_wakeUpSignal);
+}
+
+void RadioTask::sendMessage(Serializable& message) {
+    byte data[58];
+
+    // 2 Byte MessageTypeId
+    uint16_t messageTypeId = message.getMessageTypeId();
+    data[0] = static_cast<byte>(messageTypeId);
+    data[1] = static_cast<byte>(messageTypeId >> 8);
+
+    // Null the rest
+    for (uint8_t i=2; i<58; i++) data[i] = 0;
+    
+    // Serialize
+    message.serialize(data+2);
```

Comment:
Do we have any overflow protection? What if someone creates a Serializable that tries to add more than 56 bytes?